services/121-service/src/payments/fsp-integration/safaricom/services/safaricom.api.service.ts

```diff
@@ -4,67 +4,59 @@ import { TokenSet } from 'openid-client';
 
 @Injectable()
 export class SafaricomApiService {
-  public tokenSet: TokenSet;
+  private tokenSet: TokenSet;
 
-  public constructor(private readonly httpService: CustomHttpService) {}
+  public constructor(
+    private readonly httpService: CustomHttpService,
+    private readonly tokenValidationService: TokenValidationService,
+    private readonly safaricomApiHelperService: SafaricomApiHelperService,
+  ) {}
 
+  // ##TODO: Info to team: parameters were not typed here, implicit any!
   public async transfer({
     transferAmount,
     phoneNumber,
     idNumber,
     originatorConversationId,
+  }: {
+    transferAmount: number;
+    phoneNumber: string;
+    idNumber: string;
+    originatorConversationId: string;
   }): Promise<TransferReturnType> {
-    const payload = this.createTransferPayload({
+    const payload = this.safaricomApiHelperService.createTransferPayload({
       transferAmount,
       phoneNumber,
       idNumber,
       originatorConversationId,
     });
     const transferResponse = await this.makeTransferCall(payload);
 
-    let errorMessage: string | undefined;
-
-    if (!transferResponse || !transferResponse.data) {
-      errorMessage = `Error: No response data from Safaricom API`;
-    } else if (transferResponse.data.errorCode) {
-      if (transferResponse.data.errorCode === '500.002.1001') {
-        // This happens only in case of unintended Redis job re-attempt, and only if the API-request already went through the first time
-        // Return custom error, as it should be handled differently than other errors
-        const duplicateOriginatorConversationIdErrorMessage = `Error: ${transferResponse.data.errorMessage} for originatorConversationId ${originatorConversationId}`;
-        throw new DuplicateOriginatorConversationIdError(
-          duplicateOriginatorConversationIdErrorMessage,
-        );
-      }
-      errorMessage = `${transferResponse.data.errorCode} - ${transferResponse.data.errorMessage}`;
-    } else if (!transferResponse.data.ResponseCode) {
-      errorMessage = `Error: ${transferResponse.data?.statusCode} ${transferResponse.data?.error}`;
-    } else if (transferResponse.data.ResponseCode !== '0') {
-      errorMessage = `Response: ${transferResponse.data?.ResponseCode} - ${transferResponse.data?.ResponseDescription}`;
-    }
+    const errorMessage =
+      this.safaricomApiHelperService.createErrorMessageIfApplicable(
+        transferResponse,
+        originatorConversationId,
+      );
 
     if (errorMessage) {
       throw new SafaricomApiError(errorMessage);
     }
 
-    // All the checks above mean that at this stage transferResponse.data.ResponseCode === '0'
     return {
       mpesaConversationId: transferResponse.data.ConversationID,
     };
   }
 
   private async authenticate(): Promise<void> {
-    if (this.isTokenValid(this.tokenSet)) {
+    if (this.tokenValidationService.isTokenValid(this.tokenSet)) {
```

Comment:
Could we instead call the transfer function twice in the unit test? So the first time to make sure the token is valid and the second time to do the real test? I am sure you considered this, so maybe I am missing a caveat here.

(this is also more like what happens in production: first time token created, then not anymore until expired)

Reply:
For me this feels not how a unit test should work. A unit test should test a specific method with some input and some expected output. It should not initiate a flow. This relates back to [this unresolved comment ](https://github.com/global-121/121-platform/pull/6303#discussion_r1901876289) so let's continue the convo there and resolve this one.

Reply:
@RubenGeo see above discussion. After my last comment, @diderikvw and I discussed in a call that if we frame it such that the first call of the transfer method is part of the Arrange step, and the second is part of the Act step, then it might be OK. I coded it now as such. See [this commit](https://github.com/global-121/121-platform/pull/6303/commits/c024c3a7a037b53a8f32ec0b69b7f87b04b422cd). We thought it be good to get your opinin as well though.

Reply:
@jannisvisser Sounds like a good solution. Definitely better than covering this with an API test